File: string-algorithms/string_hashing.py
"""
String Hashing Functions and Utilities
======================================

Comprehensive collection of string hashing algorithms for pattern matching,
similarity detection, and data structures.

Algorithms included:
1. Polynomial Rolling Hash (Rabin-Karp)
2. Multiple Hash Functions (for collision reduction)
3. Double Hashing
4. Zobrist Hashing for strings
5. Custom hash functions with different bases and moduli

Applications:
- Fast substring matching (Rabin-Karp algorithm)
- String comparison in O(1) after O(n) preprocessing
- Hash tables for strings
- Bloom filters
- String deduplication

Time Complexity:
- Single hash: O(n) to compute, O(1) to compare
- Rolling hash: O(1) per character update
- Multiple hashes: O(k) where k = number of hash functions

Space Complexity: O(1) per hash value
"""

import random
from typing import List, Tuple, Optional
import time


class StringHash:
    """
    Base class for string hashing algorithms.
    """

    @staticmethod
    def hash_string(s: str) -> int:
        """
        Simple polynomial hash function.

        Time: O(n)
        """
        hash_val = 0
        base = 31
        mod = 10**9 + 9

        for char in s:
            hash_val = (hash_val * base + ord(char)) % mod

        return hash_val

    @staticmethod
    def hash_substring(s: str, start: int, end: int) -> int:
        """
        Hash a substring s[start:end].

        Time: O(end - start)
        """
        hash_val = 0
        base = 31
        mod = 10**9 + 9

        for i in range(start, end):
            hash_val = (hash_val * base + ord(s[i])) % mod

        return hash_val


class RollingHash:
    """
    Rolling hash (Rabin-Karp) for efficient substring hashing.

    Allows O(1) hash updates when sliding a window through text.
    """

    def __init__(self, base: int = 31, mod: int = 10**9 + 9):
        """
        Initialize rolling hash parameters.

        Args:
            base: Base for polynomial hash
            mod: Modulus for hash values
        """
        self.base = base
        self.mod = mod
        self.base_pow = {}  # Cache for base powers

    def _get_base_pow(self, n: int) -> int:
        """Get base^n mod mod with caching."""
        if n not in self.base_pow:
            self.base_pow[n] = pow(self.base, n, self.mod)
        return self.base_pow[n]

    def hash_string(self, s: str) -> int:
        """
        Compute hash of entire string.

        Time: O(n)
        """
        hash_val = 0
        for char in s:
            hash_val = (hash_val * self.base + ord(char)) % self.mod
        return hash_val

    def roll_forward(self, old_hash: int, old_char: str, new_char: str,
                    window_size: int) -> int:
        """
        Update hash when sliding window forward.

        Remove old_char from left, add new_char to right.

        Time: O(1)
        """
        # Remove contribution of old_char
        hash_val = (old_hash - ord(old_char) * \
                   self._get_base_pow(window_size - 1)) % self.mod

        # Add new_char
        hash_val = (hash_val * self.base + ord(new_char)) % self.mod

        return hash_val

    def search_pattern(self, text: str, pattern: str) -> List[int]:
        """
        Search for pattern in text using rolling hash.

        Time: O(n + m) average, O(nm) worst case (with verification)

        Returns:
            List of starting positions where pattern occurs
        """
        n = len(text)
        m = len(pattern)

        if m > n:
            return []

        matches = []

        # Compute pattern hash
        pattern_hash = self.hash_string(pattern)

        # Compute hash of first window
        window_hash = self.hash_string(text[:m])

        # Check first window
        if window_hash == pattern_hash and text[:m] == pattern:
            matches.append(0)

        # Roll through text
        for i in range(1, n - m + 1):
            window_hash = self.roll_forward(
                window_hash,
                text[i - 1],
                text[i + m - 1],
                m
            )

            # Check if hashes match
            if window_hash == pattern_hash:
                # Verify actual substring to handle collisions
                if text[i:i + m] == pattern:
                    matches.append(i)

        return matches


class MultiHash:
    """
    Multiple hash functions to reduce collision probability.

    Uses k different hash functions with different bases.
    """

    def __init__(self, k: int = 3,
                bases: Optional[List[int]] = None,
                mods: Optional[List[int]] = None):
        """
        Initialize with k hash functions.

        Args:
            k: Number of hash functions
            bases: List of bases (generated if None)
            mods: List of moduli (generated if None)
        """
        self.k = k

        if bases is None:
            self.bases = [31, 37, 41, 43, 47, 53, 59, 61][:k]
        else:
            self.bases = bases

        if mods is None:
            self.mods = [
                10**9 + 7,
                10**9 + 9,
                10**9 + 21,
                10**9 + 33,
                10**9 + 87,
                10**9 + 93,
            ][:k]
        else:
            self.mods = mods

    def hash_string(self, s: str) -> List[int]:
        """
        Compute k hash values for string.

        Time: O(kn)

        Returns:
            List of k hash values
        """
        hashes = []

        for i in range(self.k):
            hash_val = 0
            base = self.bases[i]
            mod = self.mods[i]

            for char in s:
                hash_val = (hash_val * base + ord(char)) % mod

            hashes.append(hash_val)

        return hashes

    def compare(self, hashes1: List[int], hashes2: List[int]) -> bool:
        """
        Compare two sets of hash values.

        Time: O(k)

        Returns:
            True if all hashes match
        """
        return hashes1 == hashes2


class DoubleHash:
    """
    Double hashing for even better collision resistance.

    Uses two independent hash functions.
    """

    def __init__(self):
        self.base1 = 31
        self.mod1 = 10**9 + 7
        self.base2 = 37
        self.mod2 = 10**9 + 9

    def hash_string(self, s: str) -> Tuple[int, int]:
        """
        Compute both hash values.

        Time: O(n)

        Returns:
            Tuple of (hash1, hash2)
        """
        hash1 = 0
        hash2 = 0

        for char in s:
            hash1 = (hash1 * self.base1 + ord(char)) % self.mod1
            hash2 = (hash2 * self.base2 + ord(char)) % self.mod2

        return (hash1, hash2)

    def compare(self, h1: Tuple[int, int], h2: Tuple[int, int]) -> bool:
        """
        Compare two double hash values.

        Time: O(1)
        """
        return h1 == h2


class ZobristHash:
    """
    Zobrist hashing for strings (adapted from chess).

    Uses precomputed random values for each character at each position.
    """

    def __init__(self, max_length: int = 100, seed: Optional[int] = None):
        """
        Initialize Zobrist hash tables.

        Args:
            max_length: Maximum string length to support
            seed: Random seed for reproducibility
        """
        if seed is not None:
            random.seed(seed)

        self.max_length = max_length
        self.zobrist_table = {}

        # Generate random values for each (position, character) pair
        for pos in range(max_length):
            for c in range(256):  # ASCII characters
                self.zobrist_table[(pos, c)] = random.getrandbits(64)

    def hash_string(self, s: str) -> int:
        """
        Compute Zobrist hash.

        Time: O(n)
        """
        hash_val = 0

        for i, char in enumerate(s):
            if i >= self.max_length:
                break
            hash_val ^= self.zobrist_table[(i, ord(char))]

        return hash_val

    def update_character(self, old_hash: int, pos: int,
                        old_char: str, new_char: str) -> int:
        """
        Update hash when changing a character.

        Time: O(1)
        """
        # XOR out old character
        hash_val = old_hash ^ self.zobrist_table[(pos, ord(old_char))]

        # XOR in new character
        hash_val ^= self.zobrist_table[(pos, ord(new_char))]

        return hash_val


class PrecomputedHash:
    """
    Precompute all substring hashes for O(1) queries.

    Uses prefix hash array and base powers.
    """

    def __init__(self, text: str, base: int = 31, mod: int = 10**9 + 9):
        """
        Precompute hash values.

        Time: O(n)
        """
        self.text = text
        self.n = len(text)
        self.base = base
        self.mod = mod

        # Precompute prefix hashes
        self.prefix_hash = [0] * (self.n + 1)
        for i in range(self.n):
            self.prefix_hash[i + 1] = \
                (self.prefix_hash[i] * base + ord(text[i])) % mod

        # Precompute base powers
        self.base_pow = [1] * (self.n + 1)
        for i in range(self.n):
            self.base_pow[i + 1] = (self.base_pow[i] * base) % mod

    def substring_hash(self, start: int, end: int) -> int:
        """
        Get hash of substring text[start:end] in O(1).

        Time: O(1)
        """
        hash_val = (self.prefix_hash[end] -
                   self.prefix_hash[start] * self.base_pow[end - start]) % \
                   self.mod

        return hash_val

    def compare_substrings(self, start1: int, end1: int,
                          start2: int, end2: int) -> bool:
        """
        Compare two substrings in O(1).

        Time: O(1)

        Returns:
            True if substrings are equal
        """
        if end1 - start1 != end2 - start2:
            return False

        return self.substring_hash(start1, end1) == \
               self.substring_hash(start2, end2)


# ============================================================================
# APPLICATIONS AND UTILITIES
# ============================================================================

def find_duplicate_substrings(text: str, length: int) -> List[str]:
    """
    Find all duplicate substrings of given length using hashing.

    Time: O(n) average

    Args:
        text: Input text
        length: Substring length

    Returns:
        List of duplicate substrings
    """
    if length > len(text):
        return []

    seen = {}
    duplicates = set()
    rh = RollingHash()

    # Process all substrings
    for i in range(len(text) - length + 1):
        substring = text[i:i + length]
        hash_val = rh.hash_string(substring)

        if hash_val in seen:
            duplicates.add(substring)
        else:
            seen[hash_val] = substring

    return list(duplicates)


def longest_common_prefix_hash(s1: str, s2: str) -> int:
    """
    Find length of longest common prefix using binary search and hashing.

    Time: O(log n * log n) = O(log² n)

    Returns:
        Length of longest common prefix
    """
    n = min(len(s1), len(s2))

    if n == 0:
        return 0

    # Binary search on length
    left, right = 0, n

    rh = RollingHash()

    while left < right:
        mid = (left + right + 1) // 2

        hash1 = rh.hash_string(s1[:mid])
        hash2 = rh.hash_string(s2[:mid])

        if hash1 == hash2 and s1[:mid] == s2[:mid]:
            left = mid
        else:
            right = mid - 1

    return left


# ============================================================================
# BENCHMARKING AND TESTING
# ============================================================================

def benchmark_hash_functions():
    """Benchmark different hash functions."""
    print("=" * 70)
    print("Hash Function Performance Comparison")
    print("=" * 70)

    texts = [
        "short",
        "medium length text for testing",
        "a" * 1000,  # Long repetitive
        "The quick brown fox jumps over the lazy dog" * 20
    ]

    for text in texts:
        print(f"\nText length: {len(text)}")
        print("-" * 50)

        # Simple hash
        start = time.perf_counter()
        for _ in range(1000):
            StringHash.hash_string(text)
        elapsed = (time.perf_counter() - start) * 1000
        print(f"Simple hash:    {elapsed:8.4f} ms (1000 iterations)")

        # Rolling hash
        rh = RollingHash()
        start = time.perf_counter()
        for _ in range(1000):
            rh.hash_string(text)
        elapsed = (time.perf_counter() - start) * 1000
        print(f"Rolling hash:   {elapsed:8.4f} ms (1000 iterations)")

        # Multi-hash
        mh = MultiHash(k=3)
        start = time.perf_counter()
        for _ in range(1000):
            mh.hash_string(text)
        elapsed = (time.perf_counter() - start) * 1000
        print(f"Multi-hash (3): {elapsed:8.4f} ms (1000 iterations)")

    print()


# ============================================================================
# EXAMPLES
# ============================================================================

def example_rolling_hash():
    print("=" * 70)
    print("EXAMPLE 1: Rolling Hash Pattern Matching")
    print("=" * 70)

    text = "the quick brown fox jumps over the lazy dog"
    pattern = "the"

    rh = RollingHash()
    matches = rh.search_pattern(text, pattern)

    print(f"Text: '{text}'")
    print(f"Pattern: '{pattern}'")
    print(f"Matches at positions: {matches}")

    for pos in matches:
        print(f"  Position {pos}: '{text[pos:pos+len(pattern)]}'")

    print()


def example_multi_hash():
    print("=" * 70)
    print("EXAMPLE 2: Multi-Hash Comparison")
    print("=" * 70)

    strings = ["hello", "hello", "world", "hello!", "helo"]

    mh = MultiHash(k=3)

    print("Computing hashes for strings:")
    hashes = {}
    for s in strings:
        h = mh.hash_string(s)
        hashes[s] = h
        print(f"  '{s}': {h}")

    print("\nComparing strings:")
    print(f"  'hello' == 'hello': {mh.compare(hashes['hello'], hashes['hello'])}")
    print(f"  'hello' == 'world': {mh.compare(hashes['hello'], hashes['world'])}")
    print(f"  'hello' == 'hello!': {mh.compare(hashes['hello'], hashes['hello!'])}")

    print()


def example_precomputed_hash():
    print("=" * 70)
    print("EXAMPLE 3: Precomputed Hash for Substring Queries")
    print("=" * 70)

    text = "abracadabra"
    ph = PrecomputedHash(text)

    print(f"Text: '{text}'")
    print("\nSubstring hashes (O(1) queries):")

    substrings = [(0, 4), (4, 7), (7, 11), (0, 3), (7, 10)]

    for start, end in substrings:
        hash_val = ph.substring_hash(start, end)
        print(f"  text[{start}:{end}] = '{text[start:end]}' -> {hash_val}")

    print("\nComparing substrings:")
    print(f"  text[0:3] == text[7:10]: {ph.compare_substrings(0, 3, 7, 10)}")

    print()


def example_find_duplicates():
    print("=" * 70)
    print("EXAMPLE 4: Find Duplicate Substrings")
    print("=" * 70)

    text = "banana republic banana split"
    length = 3

    duplicates = find_duplicate_substrings(text, length)

    print(f"Text: '{text}'")
    print(f"Looking for duplicates of length {length}")
    print(f"Found {len(duplicates)} duplicate substrings:")

    for dup in sorted(duplicates):
        print(f"  '{dup}'")

    print()


if __name__ == "__main__":
    print("=" * 70)
    print("STRING HASHING UTILITIES")
    print("=" * 70)
    print()

    example_rolling_hash()
    example_multi_hash()
    example_precomputed_hash()
    example_find_duplicates()
    benchmark_hash_functions()

    print("=" * 70)
    print("All examples completed!")
    print("=" * 70)

File: string-algorithms/string_hashing.go
// String Hashing Algorithms Implementation in Go
//
// String hashing is a technique to convert strings into fixed-size hash values
// for efficient comparison, pattern matching, and data structure operations.
//
// Time Complexity:
// - Hash computation: O(n) where n is string length
// - Hash comparison: O(1)
// - Rolling hash update: O(1)
//
// Go features:
// - Multiple hashing algorithms (polynomial, FNV, djb2, xxHash-inspired)
// - Rolling hash for efficient substring matching
// - Collision detection and handling
// - Thread-safe hash map
// - Rabin-Karp string matching using rolling hash
// - Concurrent hash computation with goroutines

package main

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"hash/fnv"
	"math"
	"math/big"
	"strings"
	"sync"
)

// POLYNOMIAL ROLLING HASH

const (
	DefaultBase = 31       // Common base for string hashing
	DefaultMod  = 1e9 + 9  // Large prime modulo
	Base256     = 256      // For byte-level hashing
)

// PolynomialHash computes polynomial rolling hash
//
// hash(s) = s[0]*base^(n-1) + s[1]*base^(n-2) + ... + s[n-1] (mod prime)
//
// Time Complexity: O(n)
func PolynomialHash(s string, base, mod int64) int64 {
	hash := int64(0)
	power := int64(1)

	for i := len(s) - 1; i >= 0; i-- {
		hash = (hash + int64(s[i])*power) % mod
		power = (power * base) % mod
	}

	return hash
}

// RollingHash maintains a rolling hash for efficient updates
type RollingHash struct {
	hash  int64
	base  int64
	mod   int64
	power int64 // base^(len-1) mod mod
	text  string
	start int
	end   int
}

// NewRollingHash creates a new rolling hash for a substring
func NewRollingHash(text string, start, end int, base, mod int64) *RollingHash {
	if start < 0 || end > len(text) || start >= end {
		return nil
	}

	rh := &RollingHash{
		hash:  0,
		base:  base,
		mod:   mod,
		power: 1,
		text:  text,
		start: start,
		end:   end,
	}

	// Compute initial hash and power
	length := end - start
	for i := start; i < end; i++ {
		rh.hash = (rh.hash*base + int64(text[i])) % mod
		if i < end-1 {
			rh.power = (rh.power * base) % mod
		}
	}

	_ = length
	return rh
}

// Hash returns the current hash value
func (rh *RollingHash) Hash() int64 {
	return rh.hash
}

// RollRight moves the window one position to the right
//
// Time Complexity: O(1)
func (rh *RollingHash) RollRight() bool {
	if rh.end >= len(rh.text) {
		return false
	}

	// Remove leftmost character
	rh.hash = (rh.hash - int64(rh.text[rh.start])*rh.power) % rh.mod
	if rh.hash < 0 {
		rh.hash += rh.mod
	}

	// Shift and add rightmost character
	rh.hash = (rh.hash*rh.base + int64(rh.text[rh.end])) % rh.mod

	rh.start++
	rh.end++

	return true
}

// Substring returns the current substring
func (rh *RollingHash) Substring() string {
	return rh.text[rh.start:rh.end]
}

// RABIN-KARP STRING MATCHING

// RabinKarp finds all occurrences of pattern in text using rolling hash
//
// Time Complexity: O(n+m) average case, O(nm) worst case
// where n = len(text), m = len(pattern)
func RabinKarp(text, pattern string) []int {
	if len(pattern) == 0 || len(pattern) > len(text) {
		return []int{}
	}

	const base = DefaultBase
	const mod = DefaultMod

	matches := make([]int, 0)

	// Compute pattern hash
	patternHash := PolynomialHash(pattern, base, mod)

	// Create rolling hash for text
	rh := NewRollingHash(text, 0, len(pattern), base, mod)

	// Check first window
	if rh.Hash() == patternHash && rh.Substring() == pattern {
		matches = append(matches, 0)
	}

	// Roll through text
	for rh.RollRight() {
		if rh.Hash() == patternHash && rh.Substring() == pattern {
			matches = append(matches, rh.start)
		}
	}

	return matches
}

// CLASSIC HASH FUNCTIONS

// DJB2Hash implements the djb2 hash function
//
// Simple and effective hash function by Dan Bernstein
func DJB2Hash(s string) uint64 {
	hash := uint64(5381)

	for i := 0; i < len(s); i++ {
		hash = ((hash << 5) + hash) + uint64(s[i]) // hash * 33 + c
	}

	return hash
}

// SDBMHash implements the SDBM hash function
//
// Used in SDBM database library
func SDBMHash(s string) uint64 {
	hash := uint64(0)

	for i := 0; i < len(s); i++ {
		hash = uint64(s[i]) + (hash << 6) + (hash << 16) - hash
	}

	return hash
}

// FNVHash implements FNV-1a hash
//
// Fast, well-distributed hash function
func FNVHash(s string) uint64 {
	h := fnv.New64a()
	h.Write([]byte(s))
	return h.Sum64()
}

// MurmurHash implements a simple MurmurHash-inspired function
func MurmurHash(s string) uint64 {
	const (
		seed = uint64(0x9747b28c)
		m    = uint64(0xc6a4a7935bd1e995)
		r    = 47
	)

	hash := seed ^ (uint64(len(s)) * m)

	for i := 0; i < len(s); i++ {
		k := uint64(s[i])
		k *= m
		k ^= k >> r
		k *= m

		hash ^= k
		hash *= m
	}

	hash ^= hash >> r
	hash *= m
	hash ^= hash >> r

	return hash
}

// CRYPTOGRAPHIC HASHING

// SHA256Hash computes SHA-256 hash
func SHA256Hash(s string) string {
	hash := sha256.Sum256([]byte(s))
	return hex.EncodeToString(hash[:])
}

// HASH COMPARISON AND COLLISION DETECTION

// HashComparison compares different hash functions
type HashComparison struct {
	String       string
	Polynomial   int64
	DJB2         uint64
	SDBM         uint64
	FNV          uint64
	Murmur       uint64
	SHA256       string
}

// CompareHashes computes all hash functions for a string
func CompareHashes(s string) HashComparison {
	return HashComparison{
		String:     s,
		Polynomial: PolynomialHash(s, DefaultBase, DefaultMod),
		DJB2:       DJB2Hash(s),
		SDBM:       SDBMHash(s),
		FNV:        FNVHash(s),
		Murmur:     MurmurHash(s),
		SHA256:     SHA256Hash(s)[:16], // Truncate for display
	}
}

// CollisionDetector detects hash collisions
type CollisionDetector struct {
	hashes map[int64][]string
	count  int
}

// NewCollisionDetector creates a new collision detector
func NewCollisionDetector() *CollisionDetector {
	return &CollisionDetector{
		hashes: make(map[int64][]string),
		count:  0,
	}
}

// Add adds a string and checks for collisions
func (cd *CollisionDetector) Add(s string) []string {
	hash := PolynomialHash(s, DefaultBase, DefaultMod)
	cd.count++

	if existing, ok := cd.hashes[hash]; ok {
		// Collision detected
		cd.hashes[hash] = append(existing, s)
		return existing
	}

	cd.hashes[hash] = []string{s}
	return nil
}

// Collisions returns all collision groups
func (cd *CollisionDetector) Collisions() map[int64][]string {
	collisions := make(map[int64][]string)

	for hash, strings := range cd.hashes {
		if len(strings) > 1 {
			collisions[hash] = strings
		}
	}

	return collisions
}

// LoadFactor returns the load factor
func (cd *CollisionDetector) LoadFactor() float64 {
	return float64(cd.count) / float64(len(cd.hashes))
}

// THREAD-SAFE HASH TABLE

// ThreadSafeHashTable is a concurrent hash table using string hashing
type ThreadSafeHashTable struct {
	buckets []bucket
	size    int
}

type bucket struct {
	mu    sync.RWMutex
	items map[int64][]keyValue
}

type keyValue struct {
	key   string
	value interface{}
}

// NewThreadSafeHashTable creates a new thread-safe hash table
func NewThreadSafeHashTable(size int) *ThreadSafeHashTable {
	ht := &ThreadSafeHashTable{
		buckets: make([]bucket, size),
		size:    size,
	}

	for i := 0; i < size; i++ {
		ht.buckets[i].items = make(map[int64][]keyValue)
	}

	return ht
}

// getBucket returns the bucket index for a key
func (ht *ThreadSafeHashTable) getBucket(key string) int {
	hash := PolynomialHash(key, DefaultBase, DefaultMod)
	return int(hash % int64(ht.size))
}

// Put inserts or updates a key-value pair
func (ht *ThreadSafeHashTable) Put(key string, value interface{}) {
	idx := ht.getBucket(key)
	bucket := &ht.buckets[idx]

	bucket.mu.Lock()
	defer bucket.mu.Unlock()

	hash := PolynomialHash(key, DefaultBase, DefaultMod)

	// Check if key exists
	if items, ok := bucket.items[hash]; ok {
		for i, item := range items {
			if item.key == key {
				items[i].value = value
				return
			}
		}
		bucket.items[hash] = append(items, keyValue{key, value})
	} else {
		bucket.items[hash] = []keyValue{{key, value}}
	}
}

// Get retrieves a value by key
func (ht *ThreadSafeHashTable) Get(key string) (interface{}, bool) {
	idx := ht.getBucket(key)
	bucket := &ht.buckets[idx]

	bucket.mu.RLock()
	defer bucket.mu.RUnlock()

	hash := PolynomialHash(key, DefaultBase, DefaultMod)

	if items, ok := bucket.items[hash]; ok {
		for _, item := range items {
			if item.key == key {
				return item.value, true
			}
		}
	}

	return nil, false
}

// Delete removes a key-value pair
func (ht *ThreadSafeHashTable) Delete(key string) bool {
	idx := ht.getBucket(key)
	bucket := &ht.buckets[idx]

	bucket.mu.Lock()
	defer bucket.mu.Unlock()

	hash := PolynomialHash(key, DefaultBase, DefaultMod)

	if items, ok := bucket.items[hash]; ok {
		for i, item := range items {
			if item.key == key {
				bucket.items[hash] = append(items[:i], items[i+1:]...)
				if len(bucket.items[hash]) == 0 {
					delete(bucket.items, hash)
				}
				return true
			}
		}
	}

	return false
}

// PARALLEL HASHING

// ParallelHash computes hashes of multiple strings concurrently
func ParallelHash(strings []string) []int64 {
	results := make([]int64, len(strings))
	var wg sync.WaitGroup

	for i, s := range strings {
		wg.Add(1)
		go func(idx int, str string) {
			defer wg.Done()
			results[idx] = PolynomialHash(str, DefaultBase, DefaultMod)
		}(i, s)
	}

	wg.Wait()
	return results
}

// SUBSTRING HASHING

// SubstringHashes computes hashes of all substrings of a given length
func SubstringHashes(text string, length int) map[int64][]int {
	if length > len(text) {
		return make(map[int64][]int)
	}

	hashToPositions := make(map[int64][]int)

	rh := NewRollingHash(text, 0, length, DefaultBase, DefaultMod)
	hashToPositions[rh.Hash()] = []int{0}

	for rh.RollRight() {
		hash := rh.Hash()
		hashToPositions[hash] = append(hashToPositions[hash], rh.start)
	}

	return hashToPositions
}

// FindDuplicateSubstrings finds all duplicate substrings of given length
func FindDuplicateSubstrings(text string, length int) []string {
	hashToPositions := SubstringHashes(text, length)

	duplicates := make(map[string]bool)

	for _, positions := range hashToPositions {
		if len(positions) > 1 {
			substring := text[positions[0] : positions[0]+length]
			duplicates[substring] = true
		}
	}

	result := make([]string, 0, len(duplicates))
	for dup := range duplicates {
		result = append(result, dup)
	}

	return result
}

// HASH DISTRIBUTION ANALYSIS

// AnalyzeHashDistribution analyzes hash distribution
func AnalyzeHashDistribution(strings []string, buckets int) map[int]int {
	distribution := make(map[int]int)

	for _, s := range strings {
		hash := PolynomialHash(s, DefaultBase, DefaultMod)
		bucket := int(hash % int64(buckets))
		distribution[bucket]++
	}

	return distribution
}

// DemonstrateStringHashing demonstrates string hashing algorithms
func DemonstrateStringHashing() {
	fmt.Println("🔐 String Hashing Algorithms in Go")
	fmt.Println(strings.Repeat("=", 80))

	// Basic hashing
	fmt.Println("\n📋 Basic Hash Functions:")
	fmt.Println(strings.Repeat("-", 80))

	testStrings := []string{"hello", "world", "golang", "hashing", "algorithm"}

	for _, s := range testStrings {
		comp := CompareHashes(s)
		fmt.Printf("\n'%s':\n", s)
		fmt.Printf("  Polynomial: %d\n", comp.Polynomial)
		fmt.Printf("  DJB2:       %d\n", comp.DJB2)
		fmt.Printf("  SDBM:       %d\n", comp.SDBM)
		fmt.Printf("  FNV:        %d\n", comp.FNV)
		fmt.Printf("  Murmur:     %d\n", comp.Murmur)
		fmt.Printf("  SHA256:     %s...\n", comp.SHA256)
	}

	// Rolling hash and Rabin-Karp
	fmt.Println("\n\n🔄 Rolling Hash and Rabin-Karp Pattern Matching:")
	fmt.Println(strings.Repeat("-", 80))

	text := "abracadabra"
	pattern := "abra"

	fmt.Printf("Text: '%s'\n", text)
	fmt.Printf("Pattern: '%s'\n", pattern)

	matches := RabinKarp(text, pattern)
	fmt.Printf("Matches found at positions: %v\n", matches)

	// Verify matches
	for _, pos := range matches {
		fmt.Printf("  Position %d: '%s'\n", pos, text[pos:pos+len(pattern)])
	}

	// Rolling hash demonstration
	fmt.Println("\n🎯 Rolling Hash Window:")
	fmt.Println(strings.Repeat("-", 80))

	windowSize := 3
	rh := NewRollingHash(text, 0, windowSize, DefaultBase, DefaultMod)

	fmt.Printf("Text: '%s', Window size: %d\n\n", text, windowSize)
	fmt.Printf("Position  Substring  Hash\n")
	fmt.Println(strings.Repeat("-", 30))

	fmt.Printf("%-9d %-10s %d\n", 0, rh.Substring(), rh.Hash())

	pos := 1
	for rh.RollRight() {
		fmt.Printf("%-9d %-10s %d\n", pos, rh.Substring(), rh.Hash())
		pos++
	}

	// Collision detection
	fmt.Println("\n\n⚠️  Collision Detection:")
	fmt.Println(strings.Repeat("-", 80))

	detector := NewCollisionDetector()
	testWords := []string{
		"hello", "world", "test", "hash", "collision",
		"golang", "programming", "algorithm", "data", "structure",
	}

	fmt.Println("Adding strings and checking for collisions:")
	for _, word := range testWords {
		collisions := detector.Add(word)
		if collisions != nil {
			fmt.Printf("  Collision! '%s' collides with: %v\n", word, collisions)
		}
	}

	fmt.Printf("\nTotal strings: %d\n", detector.count)
	fmt.Printf("Unique hashes: %d\n", len(detector.hashes))
	fmt.Printf("Load factor: %.2f\n", detector.LoadFactor())

	allCollisions := detector.Collisions()
	if len(allCollisions) > 0 {
		fmt.Println("\nCollision groups:")
		for hash, group := range allCollisions {
			fmt.Printf("  Hash %d: %v\n", hash, group)
		}
	}

	// Duplicate substrings
	fmt.Println("\n\n🔍 Finding Duplicate Substrings:")
	fmt.Println(strings.Repeat("-", 80))

	longText := "banana band stand banana hand"
	substringLength := 3

	fmt.Printf("Text: '%s'\n", longText)
	fmt.Printf("Substring length: %d\n\n", substringLength)

	duplicates := FindDuplicateSubstrings(longText, substringLength)
	fmt.Printf("Duplicate substrings: %v\n", duplicates)

	// Thread-safe hash table
	fmt.Println("\n\n🔒 Thread-Safe Hash Table:")
	fmt.Println(strings.Repeat("-", 80))

	ht := NewThreadSafeHashTable(10)

	// Concurrent insertions
	var wg sync.WaitGroup
	for i := 0; i < 100; i++ {
		wg.Add(1)
		go func(val int) {
			defer wg.Done()
			key := fmt.Sprintf("key%d", val)
			ht.Put(key, val)
		}(i)
	}

	wg.Wait()
	fmt.Println("Inserted 100 key-value pairs concurrently")

	// Retrieve some values
	testKeys := []string{"key0", "key50", "key99", "nonexistent"}
	for _, key := range testKeys {
		if val, ok := ht.Get(key); ok {
			fmt.Printf("  %s: %v\n", key, val)
		} else {
			fmt.Printf("  %s: not found\n", key)
		}
	}

	// Hash distribution analysis
	fmt.Println("\n\n📊 Hash Distribution Analysis:")
	fmt.Println(strings.Repeat("-", 80))

	numStrings := 1000
	testData := make([]string, numStrings)
	for i := 0; i < numStrings; i++ {
		testData[i] = fmt.Sprintf("string%d", i)
	}

	numBuckets := 20
	distribution := AnalyzeHashDistribution(testData, numBuckets)

	fmt.Printf("Distributed %d strings into %d buckets:\n\n", numStrings, numBuckets)

	// Calculate statistics
	min := math.MaxInt
	max := 0
	sum := 0

	for bucket := 0; bucket < numBuckets; bucket++ {
		count := distribution[bucket]
		if count < min {
			min = count
		}
		if count > max {
			max = count
		}
		sum += count
	}

	avg := float64(sum) / float64(numBuckets)

	fmt.Printf("Statistics:\n")
	fmt.Printf("  Min: %d\n", min)
	fmt.Printf("  Max: %d\n", max)
	fmt.Printf("  Avg: %.2f\n", avg)
	fmt.Printf("  Expected: %.2f\n", float64(numStrings)/float64(numBuckets))

	// Parallel hashing
	fmt.Println("\n\n🚀 Parallel Hash Computation:")
	fmt.Println(strings.Repeat("-", 80))

	largeDataset := make([]string, 10000)
	for i := 0; i < len(largeDataset); i++ {
		largeDataset[i] = fmt.Sprintf("data%d", i)
	}

	fmt.Printf("Computing hashes for %d strings in parallel...\n", len(largeDataset))

	start := big.NewInt(0)
	_ = start // Placeholder for timing

	hashes := ParallelHash(largeDataset)

	fmt.Printf("Computed %d hashes\n", len(hashes))
	fmt.Printf("Sample hashes: %v...\n", hashes[:5])
}

func main() {
	DemonstrateStringHashing()
	fmt.Println("\n✨ String hashing demonstration complete!")
}
